### PR TITLE
Improve Mu1 import

### PR DESCRIPTION
**Pull request by Jojo-Schmitz**
### Issue type

Import/Export issue

### Description with steps to reproduce

here esp. the vertical and horizontal frames and their margins, by syncing read114.cpp's readBox() with what read206.cpp's readBox() does.
Examples, Mu4 (and Mu3):
Screenshot 2024-07-18 190249

Mu1 (and Mu2 and with this PR):
image

Highly annoying to have to reset these 4 setting to default on every single score, I really should have done this PR back in 2019 when I switched to Mu3 and migrated loads of my old M1 and Mu2 scores. All done now, so too late for me ... (except for some 400 scores on the back burner, awaiting a properly working album feature)

### Supporting files, videos and screenshots

n/a

### What is the latest version of MuseScore Studio where this issue is present?

all Mu4 and Mu3

### Regression

no

### Operating system

all

### Checklist

- [X] This report follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting bugs and issues
- [X] I have verified that this issue has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar issues
- [X] I have attached all requested files and information to this report
- [X] I have attempted to identify the root problem as concisely as possible, and have used minimal reproducible examples where possible

### CODE REVIEW

File: src/engraving/rw/read206/read206.cpp [hunk 1]
```diff
@@ -3028,9 +3028,9 @@ static void readBox(Box* b, XmlReader& e, ReadContext& ctx)
     b->setBottomMargin(0.0);
     b->setTopGap(Millimetre(0.0));
     b->setBottomGap(Millimetre(0.0));
-    b->setAutoSizeEnabled(false);
-    b->setPropertyFlags(Pid::TOP_GAP, PropertyFlags::UNSTYLED);
-    b->setPropertyFlags(Pid::BOTTOM_GAP, PropertyFlags::UNSTYLED);
+    b->setAutoSizeEnabled(false); // didn't exist in Mu1
+    //b->setPropertyFlags(Pid::TOP_GAP, PropertyFlags::UNSTYLED);
+    //b->setPropertyFlags(Pid::BOTTOM_GAP, PropertyFlags::UNSTYLED);
```
Comment (cbjeukendrup) [inline src/engraving/rw/read206/read206.cpp:3033]: What is the reason for these changes?

File: src/engraving/rw/read114/read114.cpp [hunk 2]
```diff
@@ -2247,14 +2253,26 @@ static void readBox(XmlReader& e, ReadContext& ctx, Box* b)
...
             readBox(e, ctx, vb);
             b->add(vb);
+            keepMargins = true;           // in old file, box nesting used outer box margins
         } else if (!readBoxProperties(e, ctx, b)) {
             e.unknown();
         }
     }
+
+    // with .msc versions prior to 1.17, box margins were only used when nesting another box inside this box:
+    // for backward compatibility set them to 0.0 in all other cases, the Mu1 defaults of 5.0 just look horrible in Mu3 nd Mu4
```
Comment (cbjeukendrup) [inline src/engraving/rw/read114/read114.cpp:2268]: typo
```suggestion
    // for backward compatibility set them to 0.0 in all other cases, the Mu1 defaults of 5.0 just look horrible in Mu3 and Mu4
```

File: src/engraving/rw/read206/read206.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply src/engraving/rw/read206/read206.cpp:3033]: Importing Mu1 and Mu2 scores marked BOTOM_GAP of a vertical frame to not be the default setting even if it was

File: src/engraving/rw/read114/read114.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp]: oops, thanks

File: src/engraving/rw/read206/read206.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply src/engraving/rw/read206/read206.cpp:3033]: Yes, I should have mentioned that abobe, will do now

Comment (Jojo-Schmitz) [reply src/engraving/rw/read206/read206.cpp:3033]: TOP_GAP wasn't an issue, but commenting it out doesn't seem to harm either

Maybe we'd better remove rather than commenting out

**Comment by Jojo-Schmitz** [comment]
I'm puzzled why now, after fixing a typo on a comment, the vtests fail?!?

**Comment by cbjeukendrup** [comment]
Most likely because in the meantime a vtests-changing PR was merged to master, so a rebase should help

**Comment by Jojo-Schmitz** [comment]
done

File: src/engraving/rw/read206/read206.cpp [hunk 3]
```diff
@@ -3028,9 +3028,9 @@ static void readBox(Box* b, XmlReader& e, ReadContext& ctx)
     b->setBottomMargin(0.0);
```
Comment (cbjeukendrup) [inline src/engraving/rw/read206/read206.cpp:3028]: Aren't these first 4 lines redundant? They only set properties to values that are already their defaults. And these are `NOSTYLE` properties, so no style flags need to be set either, for these properties.

**Review by cbjeukendrup — COMMENTED** [review]
Oops these comments got stuck in "Pending"

File: src/engraving/rw/read206/read206.cpp (hunk 3 above)
Comment (Jojo-Schmitz) [reply src/engraving/rw/read206/read206.cpp]: You're right, they are redundant.

Comment (Jojo-Schmitz) [reply src/engraving/rw/read206/read206.cpp]: And the problem I tried to solve with those property flags stemmed from scores I first migrated from Mu1 to Mu2, so it is now solved there only

**Comment by Jojo-Schmitz** [comment]
This time a real vtest fail
mmrest-6-1 diff

File: src/engraving/rw/read114/read114.cpp [hunk 4]
```diff
@@ -2247,14 +2245,29 @@ static void readBox(XmlReader& e, ReadContext& ctx, Box* b)
...
+
+    // with .msc versions prior to 1.17, box margins were only used when nesting another box inside this box:
+    // for backward compatibility set them to 0.0 in all other cases, the Mu1 defaults of 5.0 just look horrible in Mu3 and Mu4
+
+    if (ctx.mscVersion() <= 114 && (b->isHBox() || b->isVBox()) && !keepMargins) {
+        b->setLeftMargin(0.0);
+        b->setRightMargin(0.0);
+        b->setTopMargin(0.0); // 2.0 would look closest to Mu1 and Mu2, but 0.0 is the default since Mu2
+        b->setBottomMargin(0.0); // 1.0 would look closest to Mu1 and Mu2, but 0.0 is the default since Mu2
+        if (b->bottomGap() != Spatium().toMM(7.0)) { // Default was 1.0, but above 6.0 are added for Mu1 scores
```
Comment (cbjeukendrup) [inline src/engraving/rw/read114/read114.cpp:2267]: > // Default was 1.0, but above 6.0 are added for Mu1 scores

I don't really get what you mean here...

Also, `Spatium().toMM(7.0)` looks really weird to me. It means: create an instance of the `Spatium` class (which means "zero spatium") and then convert that to millimetres, assuming that one spatium is 7.0 millimetres. That is absolutely not what you want here :)
Instead, how about `b->propertyDefault(Pid::BOTTOM_GAP)`?

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: See line 3097:
```cpp
    masterScore->style().set(Sid::frameSystemDistance, masterScore->style().styleS(Sid::frameSystemDistance) + Spatium(6.0));
```

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: I don't get what you mean about using `b->propertyDefault(Pid::BOTTOM_GAP)`, how?
`if (b->bottomGap() != b->propertyDefault(Pid::BOTTOM_GAP))` doesn't work

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: Hmm, maybe
`if (b->bottomGap() != b->propertyDefault(Pid::BOTTOM_GAP).value<Millimetre>())`

Comment (cbjeukendrup) [reply src/engraving/rw/read114/read114.cpp:2267]: Yes indeed, sorry.
I assume what you want to do is checking whether the value after reading matches the current style default, and if not, mark the prop as unstyled. Right?

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: Rather whether it was the old Mu1 default. But that got this 6.0 added automagicallly, as far as I can tell

Comment (cbjeukendrup) [reply src/engraving/rw/read114/read114.cpp:2267]: But actually, that should be redundant, because `readBoxProperties` (which calls `read400::TRead::readProperty`) already takes care of setting property flags correctly for values that don't match their default. So this entire `if` block would do nothing as far as I understand.

Comment (cbjeukendrup) [reply src/engraving/rw/read114/read114.cpp:2267]: And if it's _not_ redundant for some reason, then I wouldn't see why this should be inside the `!keepMargins` block.

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: But that ready the actuall setting so possible the old default of 1.0, and then makses it styled, That's what I want to fix

Comment (cbjeukendrup) [reply src/engraving/rw/read114/read114.cpp:2267]: Okay, I think I understand it slightly better now:
- In MS1, when `bottomGap` was at its default, it is not written to the file, of course
- Thus, during reading in MS4, it is not encountered
- Thus, it stays at the default that MS4 initialised it to, and stays STYLED
- "The default that MS4 initialised it to" means: the value from `Sid::frameSystemDistance` _at the time that the Box is instantiated_.
- At the time that the Box is instantiated, that style value is the value that was read while reading the `<Style>` tag, but the values in that tag had a different meaning for MS1. 
- Therefore, later, the `+ Spatium(6.0)` is done. But that is too late, because the Box is instantiated before that; so the Box contains the wrong default.

That's the problem, right?

In that case, my `b->propertyDefault(Pid::BOTTOM_GAP)` ain't gonna work either indeed, because at the point that that's called, the style value returned by it has not yet been updated.

I think the real solution would be to ensure that the style values are updated directly after reading them, before any EngravingItem is initialised. I.e. move those lines under `// adjust some styles` into the `else if (tag == "Style") {` block.

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: Makes sense, let's try that

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp:2267]: Hmm, actually now I'm completly confised and don't even know anymore what I really wanted to fix.
So let's just drop that one for now and restring to resetting the frame borders (by moving the code from read2006.cpp to read114.cpp and removing apparently redundant stuff)

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp]: I wonder what that does to the mtest and vtests...

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp]: OK, one mtest file was affected, and for good reason.

Actually the issue I wanted to fix was an issue with 2.x importing 1.x files, doesn't happen with 3.x nor 4.x
It does seem to happen though when importing a 2.x score into 3.x or 4.x, if that score originally stemmed from 1.x (and still had that property marked as unstyled), 2.x set the bottom gap to 7.0 (and unstyled) 3.x and 4.x set it to 8.0 (!?!) and still unstyled. So if anything this issue would need to get fixed in read206.cpp

Comment (Jojo-Schmitz) [reply src/engraving/rw/read114/read114.cpp]: If you got some spare time and are curious, check the attached
[Frametests.zip](https://github.com/user-attachments/files/16325165/Frametests.zip)

File: src/engraving/rw/read206/read206.cpp [hunk 5]
```diff
@@ -3022,32 +3022,21 @@ static void readMeasure206(Measure* m, int staffIdx, XmlReader& e, ReadContext&
...
 {
-    b->setLeftMargin(0.0);
-    b->setRightMargin(0.0);
-    b->setTopMargin(0.0);
-    b->setBottomMargin(0.0);
-    b->setTopGap(Millimetre(0.0));
-    b->setBottomGap(Millimetre(0.0));
-    b->setAutoSizeEnabled(false);
-    b->setPropertyFlags(Pid::TOP_GAP, PropertyFlags::UNSTYLED);
-    b->setPropertyFlags(Pid::BOTTOM_GAP, PropertyFlags::UNSTYLED);
```
Comment (Jojo-Schmitz) [inline src/engraving/rw/read206/read206.cpp:3033]: Actually removing the above two lines fixed that styled issue with Mu2 scores that started their live in Mu1 already.
Only they still show their bottom gap 1.0sp larger than in Mu2.
Mu1 default was 1.0, Mu2 default was 0.0 (so the import marked the Mu1 1.0 as unstyled), Mu3 and Mu4 default is 7.0 and when reading a Mu2 score the current (unstyled) value and the current default get added, at least for the top frame, see a few lines down, in `readStaffContent206()`

**Comment by Jojo-Schmitz** [comment]
Just found another problem: frame text, esp Texter and Composer (i.e. those with a bottom alignment) have a problem on Mu1 import. when their height is less than the height of those texts. Mu3 had that problem too. Culprit seems that texts that did fit their frame in Mu1 no longer do in Mu3 and Mu3 for reasons that are beyond me.
A text that in Mu1, Mu2 and M3 did fix a vertical frame with a height of 161sp (10pt Arial bold) in Mu4.3.2 needs 191sp, almost 20% more!?! Mu4.4 (with and without this PR here) needs a bit less apparently, 'only' 184.5sp, still almost 15% more than Mu1-3.
Also Mu4.3.2 looses the boldness and makes the font FreeSerif, that seems to be the reason for the extra space it needs
4.4 'only' looses the boldness, adding that back and it needs 191sp too.
Ah, now I see: it incleases the font size to 12pt, exactly those 20%.!

**Comment by Jojo-Schmitz** [comment]
I guess fixing this is for another day and another PR...

**Comment by cbjeukendrup** [comment]
I think this PR looks good now. Will double-check with the engraving team though to be sure they are okay with it too. 

I have to correct one thing I said earlier: 
`read400::TRead::readProperty` marks a property as `UNSTYLED` always when that property is encountered in a file if that property has a Sid connected to it.
So, contrary to what I said at some point, it does _not_ compare the property value from the file with any style value. The fact alone that the property had been written to the file at all, means that it had been modified and thus is `UNSTYLED`.

I don't think that information will have any further influence on this PR, but it might be helpful when solving other problems in the future.

**Comment by cbjeukendrup** [comment]
@mike-spa The summary of this PR in its final state is:
- for Mu1 scores, the box margins of frames need to be discarded for frames that contain only text (and not subframes). For unknown reasons, this code was in read206 instead of read114, and thus didn't do its job.
- besides fixing that, the PR removes some `setTopMargin(0.0)` etc. calls that are useless, and some `setTopGap` calls that turned out to do more harm than good.

I think the PR looks good but wanted to check that it doesn't have any risks that I don't see.

**Review by mike-spa — APPROVED** [review]
Don't see any problems with this, as long as it does what it is intended to and @cbjeukendrup is ok with it.

As a general consideration, may I just point out @Jojo-Schmitz that if it wasn't for Casper summary comment at the end, it would be almost impossible for me to understand what this PR is trying to do? Next times please
- Keep the PR description shorter, clearer and to the point
- Make sure that the screeshots you include aren't misleading (I still don't understand why in the second screenshot the title and subtitle have a smaller font?)
- Keep comments short and to the point (we don't need comments sayings that vtests fail, especially when you know they just need a rebase)

**Comment by Jojo-Schmitz** [comment]
> * Keep the PR description shorter, clearer and to the point
What is too long in the description, in the initial one? I don#t think it can be any shorter

Further ones are trying to explain in more detail where it seems to go wrong, parts of my analysis, trouble shooting, and development progress.

> * Make sure that the screeshots you include aren't misleading (I still don't understand why in the second screenshot the title and subtitle have a smaller font?)

The screenshots just have different resolutions, for Mu2 I used MuseScore's builtin foto-mode, for Mu4 the Windows screenshots tool, the font sizese are identical, it's their position I wanted to point out,

> * Keep comments short and to the point (we don't need comments sayings that vtests fail, especially when you know they just need a rebase)
At the time I didn't know that a rebase would be sufficient.
